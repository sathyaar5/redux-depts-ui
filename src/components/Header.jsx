import React, { useState } from "react";
import { Typography, Button, Modal, TextField } from "@mui/material";
import { useDispatch, useSelector } from "react-redux";

import { addEmployee } from "../actions";

const FIELD_NAMES = [
  { fieldName: "name", label: "Name" },
  { fieldName: "designation", label: "Designation" },
  { fieldName: "branch", label: "Branch" },
];

const Header = ({ department, numberOfEmployees, manager }) => {
  const [open, setOpen] = useState(false);
  const [employeeData, setEmployeeData] = useState({});
  const [error, setError] = useState("");
  const dispatch = useDispatch();
  const selectedDepartmentIndex = useSelector(
    (state) => state.selectedDepartmentIndex,
  );

  const handleOpen = () => setOpen(true);
  const handleClose = () => {
    setEmployeeData({});
    setError("");
    setOpen(false);
  };

  const handleInputChange = (event) => {
    const { name, value } = event.target;
    setEmployeeData({ ...employeeData, [name]: value });
  };

  const handleAddEmployee = () => {
    if (
      !employeeData.name || !employeeData.designation || !employeeData.branch
    ) {
      setError("Please fill in all the required fields.");
      return;
    }

    const newEmployeeId =
      department.employeeDetails.length > 0 ? department.employeeDetails[department.employeeDetails.length - 1].id + 1 : 1;
    const newEmployee = { id: newEmployeeId, ...employeeData };

    dispatch(addEmployee(selectedDepartmentIndex, newEmployee));
    handleClose();
  };

  return (
    <div className="department-details">
      <Typography variant="h4" gutterBottom>
        {department.deptName}
      </Typography>
      <Typography variant="body1" gutterBottom>
        No of employees: {numberOfEmployees}
      </Typography>
      <Typography variant="body1" gutterBottom>
        Manager: {manager}
      </Typography>
      <Button variant="contained" onClick={handleOpen}>
        Add Employee
      </Button>
      <Modal open={open} onClose={handleClose}>
        <div
          style={{
            position: "absolute",
            top: "50%",
            left: "50%",
            transform: "translate(-50%, -50%)",
            backgroundColor: "#fff",
            padding: "20px",
          }}
        >
          <Typography variant="h6">Add Employee</Typography>
          {FIELD_NAMES.map(({ fieldName, label }) => (
            <TextField
              key={fieldName}
              name={fieldName}
              label={label}
              value={employeeData[fieldName] || ""}
              onChange={handleInputChange}
              fullWidth
              margin="normal"
            />
          ))}
          {error && (
            <Typography color="error" variant="body2">
              {error}
            </Typography>
          )}
          <Button variant="contained" onClick={handleAddEmployee}>
            Save
          </Button>
        </div>
      </Modal>
    </div>
  );
};

export default Header;