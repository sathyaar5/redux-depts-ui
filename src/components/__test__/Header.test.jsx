import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import Header from "../Header";
import store from "../../store";

test("renders header details correctly", () => {
  const department = {
    deptName: "Technical Department",
    numberOfEmployees: 5,
    manager: "Abdul hameed",
  };

  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <Header department={department} />
      </MemoryRouter>
    </Provider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
