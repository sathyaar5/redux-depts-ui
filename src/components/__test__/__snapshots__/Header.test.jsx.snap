// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders header details correctly 1`] = `
<div
  className="department-details"
>
  <h4
    className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
  >
    Technical Department
  </h4>
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
  >
    No of employees: 
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
  >
    Manager: 
  </p>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Add Employee
  </button>
</div>
`;
